File: tp_tda/tp1/bOrW.py
import time

def bOrW(s, i, b, w, mem):
    if i == len(s):
        return 0
    
    if (i, b, w) in mem:
        return mem[(i, b, w)]
    
    unpainted = bOrW(s, i+1, b, w, mem) + 1
    
    if b == -1 or s[i] > s[b]:
        black = bOrW(s, i+1, i, w, mem)
    else:
        black = float('inf')
        
    if w == -1 or s[i] < s[w]:
        white = bOrW(s, i+1, b, i, mem)
    else:
        white = float('inf')
        
    res = min(unpainted, white, black)
    mem[(i, b, w)] = res
    return res

def main():
    n = int(input())
    while n != -1:
        sequence = input()
        sequence = list(map(int, sequence.split()))
        mem = {}
        
        inicio = time.time()
        resultado = bOrW(sequence, 0, -1, -1, mem)
        fin = time.time()
        
        print(f"Resultado: {resultado}")
        print(f"Tiempo de ejecución: {fin - inicio} segundos")
        
        n = int(input())

main()

File: tp_tda/tp1/bOrW.cpp
#include <iostream>
#include <vector>
#include <limits>
#include <algorithm>

using namespace std;

int bOrW(vector<int>& s, int i, int b, int w, vector<vector<vector<int>>>& mem) {
    if (i == s.size()) {
        return 0;
    }

    if (mem[i][b + 1][w + 1] != -1) {
        return mem[i][b + 1][w + 1];
    }

    int unpainted = bOrW(s, i + 1, b, w, mem) + 1;

    int black = numeric_limits<int>::max();
    if (b == -1 || s[i] > s[b]) {
        black = bOrW(s, i + 1, i, w, mem);
    }

    int white = numeric_limits<int>::max();
    if (w == -1 || s[i] < s[w]) {
        white = bOrW(s, i + 1, b, i, mem);
    }

    int res = min({unpainted, black, white});
    mem[i][b + 1][w + 1] = res;
    return res;
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    int n;
    cin >> n;
    while (n != -1) {
        vector<int> sequence(n);
        for (int i = 0; i < n; ++i) {
            cin >> sequence[i];
        }

        vector<vector<vector<int>>> mem(n, vector<vector<int>>(n + 1, vector<int>(n + 1, -1)));
        cout << bOrW(sequence, 0, -1, -1, mem) << endl;

        cin >> n;
    }
    return 0;
}
